File: leetcode/algo2/day6/200_number_of_islands.cpp
#include <vector>
#include <iostream>
#include <queue>
#include <utility>
int numIslands(std::vector<std::vector<char>> &grid)
{
    int m = grid.size();    // vertical;
    int n = grid[0].size(); // horizontal;
    int island_count = 0;

    std::vector<int> dir_x = {0, 1, 0, -1};
    std::vector<int> dir_y = {-1, 0, 1, 0};

    std::queue<std::pair<int, int>> visit_queue;
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (grid[i][j] == '1')
            {
                visit_queue.push(std::make_pair(i, j));
                grid[i][j] = '0';
                while (!visit_queue.empty())
                {
                    std::pair curr = visit_queue.front();
                    visit_queue.pop();
                    for (int k = 0; k < dir_x.size(); k++)
                    {
                        if (curr.first + dir_x[k] >= 0 && curr.second + dir_y[k] >= 0 && curr.first + dir_x[k] < m && curr.second + dir_y[k] < n && grid[curr.first + dir_x[k]][curr.second + dir_y[k]] != '0')
                        {
                            visit_queue.push(std::make_pair(curr.first + dir_x[k], curr.second + dir_y[k]));
                            grid[curr.first + dir_x[k]][curr.second + dir_y[k]] = '0';
                        }
                    }
                }
                island_count++;
            }
        }
    }
    return island_count;
}
int main()
{
    // std::vector<std::vector<char>> grid = {
    //     {'1', '1', '0', '0', '0'},
    //     {'1', '1', '0', '0', '0'},
    //     {'0', '0', '1', '0', '0'},
    //     {'0', '0', '0', '1', '1'}};
    std::vector<std::vector<char>> grid = {
        {'1', '1', '1', '1', '0'},
        {'1', '1', '0', '1', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '0', '0', '0'}};
    std::cout << numIslands(grid) << std::endl;
}

File: leetcode/algo2/day6/200_number_of_islands.js
var numIslands = function (grid) {
  let m = grid.length;
  let n = grid[0].length;
  let island_count = 0;

  let dir_x = [0, 1, 0, -1];
  let dir_y = [-1, 0, 1, 0];

  let visit_queue = [];

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === "1") {
        visit_queue.push([i, j]);
        grid[i][j] = "0";
        while (visit_queue.length) {
          let curr = visit_queue.shift();
          for (let k = 0; k < dir_x.length; k++) {
            if (
              curr[0] + dir_x[k] >= 0 &&
              curr[1] + dir_y[k] >= 0 &&
              curr[0] + dir_x[k] < m &&
              curr[1] + dir_y[k] < n &&
              grid[curr[0] + dir_x[k]][curr[1] + dir_y[k]] !== "0"
            ) {
              visit_queue.push([curr[0] + dir_x[k], curr[1] + dir_y[k]]);
              grid[curr[0] + dir_x[k]][curr[1] + dir_y[k]] = "0";
            }
          }
        }
        island_count++;
      }
    }
  }
  return island_count;
};
grid = [
  ["1", "1", "0", "0", "0"],
  ["1", "1", "0", "0", "0"],
  ["0", "0", "1", "0", "0"],
  ["0", "0", "0", "1", "1"],
];
console.log(numIslands(grid));
